### PR TITLE
Use separate event stream per namespace

### LINKED ISSUE #1381: Re-work event streams in blockchain plugins to create one eventstream per namespace

### PR DESCRIPTION
**Pull request by nguyer**
This PR resolves https://github.com/hyperledger/firefly/issues/1381 and https://github.com/hyperledger/firefly/issues/1382

Because we are moving from a single event stream to one per namespace, existing subscriptions will be recreated and events will be replayed on a new event stream the first time a FireFly Core node starts with these changes. Otherwise there is no user impacting change in this PR.

Event streams now follow the format `<plugin_topic_name>/<namespace_name>`. For example, event stream for the `default` namespace with a plugin topic of `0` would now be: `0/default`.

There is still outstanding work to do in the ~tokens plugin and~ tokens connectors, but I'm splitting those out into separate PRs.

### CODE REVIEW

**Comment by nguyer** [comment]
If we have multiple namespaces using the same token connector (which I think is a valid configuration) with these changes FireFly would create multiple WebSocket connections to that one token connector which may not work correctly until the token connectors themselves are updated. I'm not sure if that's an issue or not.

**Review by peterbroadhurst — CHANGES_REQUESTED** [review]
This is a great piece of work @nguyer 👍 

A few comments/requests to think through.

File: internal/blockchain/fabric/fabric.go [hunk 1]
```diff
@@ -245,28 +239,68 @@ func (f *Fabric) Init(ctx context.Context, cancelCtx context.CancelFunc, conf co
...
 
-	f.backgroundStart = f.fabconnectConf.GetBool(FabconnectBackgroundStart)
+	return nil
+}
+
+func (f *Fabric) getTopic(namespace string) string {
+	return fmt.Sprintf("%s/%s", f.pluginTopic, namespace)
+}
+
+func (f *Fabric) StartNamespace(ctx context.Context, namespace string) (err error) {
```
[the remark this answers is not in the document]
Comment (nguyer) [reply internal/blockchain/fabric/fabric.go:254]: This is a good call out. I'm investigating this...

File: internal/orchestrator/orchestrator.go [hunk 2]
```diff
@@ -249,6 +249,16 @@ func (or *orchestrator) Init() (err error) {
 func Purge(ctx context.Context, ns *core.Namespace, plugins *Plugins, dxNodeName string) {
 	// Clear all handlers on all plugins, as this namespace is never coming back
 	setHandlers(ctx, plugins, ns, dxNodeName, nil, nil)
+	err := plugins.Blockchain.Plugin.StopNamespace(ctx, ns.Name)
```
[the remark this answers is not in the document]
Comment (nguyer) [reply internal/orchestrator/orchestrator.go:252]: It's probably just in the wrong spot. I'll move it. Thanks.

[the remark this answers is not in the document]
Comment (nguyer) [reply internal/orchestrator/orchestrator.go:252]: Looking at this one again... it doesn't look like `Stop` gets called from `Purge`. Should it? Or does something else call both of them?

**Review by peterbroadhurst — APPROVED** [review]
There are a few minor additional comments from me to think about @nguyer, but none are a blocker to getting this important piece of work into mainline and progressing forwards to a release.

File: internal/blockchain/ethereum/ethereum.go [hunk 3]
```diff
@@ -199,29 +193,68 @@ func (e *Ethereum) Init(ctx context.Context, cancelCtx context.CancelFunc, conf
...
+
+	return nil
+}
 
-	e.closed = make(chan struct{})
-	go e.eventLoop()
+func (e *Ethereum) StopNamespace(ctx context.Context, namespace string) (err error) {
+	wsconn, ok := e.wsconn[namespace]
+	if ok {
+		wsconn.Close()
```
[the remark this answers is not in the document]
Comment (nguyer) [reply internal/blockchain/ethereum/ethereum.go:254]: Hmm. I added a wait for this here, but it then gets into a deadlock trying to stop the namespace
```go
if ok {
    <-e.closed[namespace]
    wsconn.Close()
}
```

File: internal/blockchain/ethereum/ethereum.go [hunk 4]
```diff
@@ -199,29 +193,68 @@ func (e *Ethereum) Init(ctx context.Context, cancelCtx context.CancelFunc, conf
...
+	err = e.wsconn[namespace].Connect()
 	if err != nil {
 		return err
 	}
 
-	e.streamID = stream.ID
-	log.L(e.ctx).Infof("Event stream: %s (topic=%s)", e.streamID, e.topic)
+	e.closed[namespace] = make(chan struct{})
+
+	go e.eventLoop(namespace, e.wsconn[namespace], e.closed[namespace])
```
[the remark this answers is not in the document]
Comment (nguyer) [reply internal/blockchain/ethereum/ethereum.go:245]: Thanks. Added below in the event loop

File: internal/blockchain/fabric/fabric.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nguyer) [reply internal/blockchain/fabric/fabric.go:254]: Noted in: https://github.com/hyperledger/firefly/pull/1461/files#diff-b2e77d2a6574b8e68136b226be7c14dff6897e0c30751d33585d9d2bb916cea9R65

**Comment by nguyer** [comment]
I am also noting that the Tezos plugin was developed in parallel to this PR so it does not have these changes. However, TezosConnect does not yet have evenstream support implemented. Once Tezosconnect supports eventstreams we should circle back and make these same changes there too. If I were to do the changes as part of this PR today I wouldn't be able to test them, so I think it's best to hold off for now.

**Pull request merged**